File: plugins/capacitor-passkey-prf/android/src/main/kotlin/technology/breez/glow/passkeyprf/PasskeyPrfPlugin.kt
package technology.breez.glow.passkeyprf

import android.app.Activity
import android.util.Base64
import android.webkit.WebView
import com.getcapacitor.JSArray
import com.getcapacitor.JSObject
import com.getcapacitor.Plugin
import com.getcapacitor.PluginCall
import com.getcapacitor.PluginMethod
import com.getcapacitor.annotation.CapacitorPlugin
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import technology.breez.spark.passkey.PasskeyProvider
import breez_sdk_spark.ConnectWithPasskeyRequest
import breez_sdk_spark.PasskeyAvailability
import breez_sdk_spark.PasskeyClient
import breez_sdk_spark.PasskeyConfig
import breez_sdk_spark.PasskeyCredential
import breez_sdk_spark.PasskeyException
import breez_sdk_spark.PasskeyProviderOptions
import breez_sdk_spark.PrfProviderException
import breez_sdk_spark.RegisterRequest
import breez_sdk_spark.Seed
import breez_sdk_spark.SignInRequest
import breez_sdk_spark.Wallet

/**
 * Capacitor bridge over `breez_sdk_spark.PasskeyClient`.
 *
 * Owns a single `PasskeyClient` per session (built in `initialize`)
 * plus a process-lifetime `BlockStoreCredentialRegistry`.
 */
@CapacitorPlugin(name = "Passkey")
class PasskeyPlugin : Plugin() {
    private val scope = CoroutineScope(Dispatchers.Main)

    private val registry: BlockStoreCredentialRegistry by lazy {
        BlockStoreCredentialRegistry(context.applicationContext)
    }

    @Volatile
    private var client: PasskeyClient? = null

    // RP the client was initialized with; needed to key the registry.
    @Volatile
    private var rpId: String? = null

    /**
     * GrapheneOS detection. The OS is fingerprint-resistant (stock-like
     * Build.* fields, stock user agent), so the WebView provider being
     * Vanadium is the one reliable native signal. Surfaced as
     * `isGrapheneOs` on the `checkAvailability()` result so the web layer
     * can show targeted copy when passkey ceremonies fail there (Google
     * Password Manager cannot verify the screen lock under sandboxed
     * Play services and times out).
     */
    private val isGrapheneOs: Boolean by lazy {
        try {
            WebView.getCurrentWebViewPackage()?.packageName
                ?.contains("vanadium", ignoreCase = true) == true
        } catch (e: Exception) {
            // Detection must never break availability reporting.
            false
        }
    }

    private companion object {
        /**
         * RP IDs this app may run a ceremony against. Must stay in step with
         * the domains in assetlinks.json (and iOS's associated-domains
         * entitlement): a domain here that the app is not associated with
         * dead-ends the OS sheet.
         */
        val ALLOWED_RP_IDS = setOf("keys.hayabusawallet.com")
    }

    @PluginMethod
    fun initialize(call: PluginCall) {
        val rpId = call.getString("rpId")
        val rpName = call.getString("rpName")
        if (rpId == null || rpName == null) {
            call.reject("rpId and rpName are required", "INVALID_ARGUMENT")
            return
        }
        // The RP is ours to decide, not the WebView's. Settle it here, before
        // a ceremony can put a domain in front of the user.
        if (rpId !in ALLOWED_RP_IDS) {
            call.reject("rpId '$rpId' is not an associated domain of this app", "INVALID_ARGUMENT")
            return
        }
        this.rpId = rpId
        val provider = PasskeyProvider(
            activityProvider = { activity as Activity },
            options = PasskeyProviderOptions(
                rpId = rpId,
                rpName = rpName,
                userName = call.getString("userName"),
                userDisplayName = call.getString("userDisplayName"),
            ),
        )
        val config = call.getString("defaultLabel")?.let { PasskeyConfig(defaultLabel = it) }
        client = PasskeyClient(provider, call.getString("breezApiKey"), config)
        call.resolve()
    }

    @PluginMethod
    fun checkAvailability(call: PluginCall) {
        val c = client ?: run {
            call.resolve(JSObject().put("type", "prfUnsupported").put("isGrapheneOs", isGrapheneOs))
            return
        }
        scope.launch {
            try {
                call.resolve(encodeAvailability(c.checkAvailability()).put("isGrapheneOs", isGrapheneOs))
            } catch (e: Exception) {
                call.reject(e.message ?: "checkAvailability failed", errorCode(e))
            }
        }
    }

    @PluginMethod
    fun register(call: PluginCall) {
        val c = requireClient(call) ?: return
        val request = RegisterRequest(
            label = call.getString("label"),
            excludeCredentials = parseB64Array(call, "excludeCredentials"),
        )
        scope.launch {
            try {
                val response = c.register(request)
                recordCredential(response.credential)
                call.resolve(JSObject().apply {
                    put("wallet", encodeWallet(response.wallet))
                    put("credential", response.credential?.let { encodeCredential(it) })
                })
            } catch (e: Exception) {
                call.reject(e.message ?: "register failed", errorCode(e))
            }
        }
    }

    @PluginMethod
    fun signIn(call: PluginCall) {
        val c = requireClient(call) ?: return
        val request = SignInRequest(
            label = call.getString("label"),
            allowCredentials = parseB64Array(call, "allowCredentials"),
            preferImmediatelyAvailableCredentials = call.getBoolean("preferImmediatelyAvailableCredentials"),
        )
        scope.launch {
            try {
                val response = c.signIn(request)
                recordCredential(response.credential)
                call.resolve(JSObject().apply {
                    put("wallet", encodeWallet(response.wallet))
                    put("labels", JSArray(response.labels))
                    put("credentialId", response.credential?.credentialId?.let { encodeB64(it) })
                })
            } catch (e: Exception) {
                call.reject(e.message ?: "signIn failed", errorCode(e))
            }
        }
    }

    @PluginMethod
    fun connectWithPasskey(call: PluginCall) {
        val c = requireClient(call) ?: return
        val request = ConnectWithPasskeyRequest(
            label = call.getString("label"),
            allowCredentials = parseB64Array(call, "allowCredentials"),
            excludeCredentials = parseB64Array(call, "excludeCredentials"),
        )
        scope.launch {
            try {
                val response = c.connectWithPasskey(request)
                recordCredential(response.credential)
                call.resolve(JSObject().apply {
                    put("wallet", encodeWallet(response.wallet))
                    // Discovery labels for the returning multi-wallet case;
                    // empty on the register path (glow-web#309).
                    put("labels", JSArray(response.labels))
                    put("registeredCredential", response.credential?.let { encodeCredential(it) })
                })
            } catch (e: Exception) {
                call.reject(e.message ?: "connectWithPasskey failed", errorCode(e))
            }
        }
    }

    @PluginMethod
    fun listLabels(call: PluginCall) {
        val c = requireClient(call) ?: return
        scope.launch {
            try {
                call.resolve(JSObject().put("labels", JSArray(c.labels().list())))
            } catch (e: Exception) {
                call.reject(e.message ?: "listLabels failed", errorCode(e))
            }
        }
    }

    @PluginMethod
    fun storeLabel(call: PluginCall) {
        val label = call.getString("label") ?: run {
            call.reject("Missing required parameter: label", "INVALID_ARGUMENT")
            return
        }
        val c = requireClient(call) ?: return
        scope.launch {
            try {
                c.labels().store(label)
                call.resolve()
            } catch (e: Exception) {
                call.reject(e.message ?: "storeLabel failed", errorCode(e))
            }
        }
    }

    @PluginMethod
    fun getKnownCredentialIds(call: PluginCall) {
        val rp = rpId ?: run {
            call.resolve(JSObject().put("credentialIds", JSArray()))
            return
        }
        scope.launch {
            try {
                val ids = registry.read(rp)
                val arr = JSArray().apply { ids.forEach { put(encodeB64(it)) } }
                call.resolve(JSObject().put("credentialIds", arr))
            } catch (e: Exception) {
                call.reject(e.message ?: "getKnownCredentialIds failed", errorCode(e))
            }
        }
    }

    @PluginMethod
    fun removeKnownCredentialId(call: PluginCall) {
        val raw = call.getString("credentialId")
        if (raw.isNullOrEmpty()) {
            call.reject("Missing required parameter: credentialId", "INVALID_ARGUMENT")
            return
        }
        val decoded = decodeB64Either(raw)
        if (decoded == null) {
            call.reject("credentialId is not valid base64", "INVALID_ARGUMENT")
            return
        }
        val rp = rpId ?: run {
            call.resolve()
            return
        }
        scope.launch {
            try {
                registry.remove(rp, decoded)
                call.resolve()
            } catch (e: Exception) {
                call.reject(e.message ?: "removeKnownCredentialId failed", errorCode(e))
            }
        }
    }

    @PluginMethod
    fun clearKnownCredentialIds(call: PluginCall) {
        val rp = rpId ?: run {
            call.resolve()
            return
        }
        scope.launch {
            try {
                registry.clear(rp)
                call.resolve()
            } catch (e: Exception) {
                call.reject(e.message ?: "clearKnownCredentialIds failed", errorCode(e))
            }
        }
    }

    /**
     * Best-effort: record a credential ID in the synced store.
     */
    private suspend fun recordCredential(cred: PasskeyCredential?) {
        val rp = rpId ?: return
        if (cred == null) return
        try {
            registry.add(rp, cred.credentialId)
        } catch (e: Exception) {
            // Best-effort; a registry write failure must not fail the ceremony.
        }
    }

    private fun requireClient(call: PluginCall): PasskeyClient? {
        val c = client
        if (c == null) call.reject("Plugin not initialized", "NOT_INITIALIZED")
        return c
    }

    private fun parseB64Array(call: PluginCall, key: String): List<ByteArray> {
        val raw = call.getArray(key)?.toList<String>() ?: return emptyList()
        return raw.mapNotNull { decodeB64Either(it) }
    }

    private fun encodeB64(bytes: ByteArray): String =
        Base64.encodeToString(bytes, Base64.NO_WRAP)

    private fun decodeB64Either(input: String): ByteArray? {
        listOf(
            Base64.URL_SAFE or Base64.NO_WRAP or Base64.NO_PADDING,
            Base64.URL_SAFE or Base64.NO_WRAP,
            Base64.NO_WRAP,
            Base64.NO_WRAP or Base64.NO_PADDING,
        ).forEach { flags ->
            try { return Base64.decode(input, flags) }
            catch (_: IllegalArgumentException) { /* try next */ }
        }
        return null
    }

    private fun encodeWallet(wallet: Wallet): JSObject =
        JSObject().apply {
            put("seed", encodeSeed(wallet.seed))
            put("label", wallet.label)
        }

    private fun encodeSeed(seed: Seed): JSObject = when (seed) {
        is Seed.Mnemonic -> JSObject().apply {
            put("type", "mnemonic")
            put("mnemonic", seed.mnemonic)
            put("passphrase", seed.passphrase)
        }
        is Seed.Entropy -> JSObject().apply {
            put("type", "entropy")
            put("entropy", encodeB64(seed.v1))
        }
    }

    private fun encodeCredential(cred: PasskeyCredential): JSObject =
        JSObject().apply {
            put("credentialId", encodeB64(cred.credentialId))
            put("userId", cred.userId?.let { encodeB64(it) })
            put("aaguid", cred.aaguid?.let { encodeB64(it) })
            put("backupEligible", cred.backupEligible)
        }

    private fun encodeAvailability(a: PasskeyAvailability): JSObject = when (a) {
        is PasskeyAvailability.Available ->
            JSObject().put("type", "available")
        is PasskeyAvailability.PrfUnsupported ->
            JSObject().put("type", "prfUnsupported")
        is PasskeyAvailability.NotAssociated ->
            JSObject().put("type", "notAssociated").put("source", a.source).put("reason", a.reason)
        is PasskeyAvailability.Skipped ->
            JSObject().put("type", "skipped").put("reason", a.reason)
    }

    private fun errorCode(e: Exception): String = when (e) {
        is PasskeyException -> passkeyErrorCode(e)
        is PrfProviderException -> prfErrorCode(e)
        else -> "UNKNOWN_ERROR"
    }

    private fun passkeyErrorCode(e: PasskeyException): String = when (e) {
        // The SDK wraps every PrfProviderException in here, and this arm is
        // matched before the bare `is PrfProviderException` one in errorCode,
        // so unwrap: collapsing to GENERIC_ERROR loses the no-credential,
        // cancelled, and already-exists cases the JS routing depends on.
        is PasskeyException.Prf -> prfErrorCode(e.v1)
        is PasskeyException.RelayConnectionFailed -> "RELAY_CONNECTION_FAILED"
        is PasskeyException.NostrWriteFailed -> "NOSTR_WRITE_FAILED"
        is PasskeyException.NostrReadFailed -> "NOSTR_READ_FAILED"
        is PasskeyException.KeyDerivationException -> "KEY_DERIVATION_ERROR"
        is PasskeyException.InvalidPrfOutput -> "INVALID_PRF_OUTPUT"
        is PasskeyException.MnemonicException -> "MNEMONIC_ERROR"
        is PasskeyException.InvalidSalt -> "INVALID_SALT"
        // A passkey exists on the device even though the ceremony failed.
        // Its own code so the web layer signs in with it rather than
        // offering a create that would strand it.
        is PasskeyException.CreatedButNotDerived -> "CREATED_BUT_NOT_DERIVED"
        is PasskeyException.Generic -> "GENERIC_ERROR"
    }

    private fun prfErrorCode(e: PrfProviderException): String = when (e) {
        is PrfProviderException.PrfNotSupported -> "PRF_NOT_SUPPORTED"
        is PrfProviderException.UserCancelled -> "USER_CANCELLED"
        is PrfProviderException.UserTimedOut -> "USER_TIMED_OUT"
        is PrfProviderException.CredentialNotFound -> "CREDENTIAL_NOT_FOUND"
        is PrfProviderException.AuthenticationFailed -> "AUTHENTICATION_FAILED"
        is PrfProviderException.PrfEvaluationFailed -> "PRF_EVALUATION_FAILED"
        is PrfProviderException.Configuration -> "CONFIGURATION_ERROR"
        is PrfProviderException.CredentialAlreadyExists -> "CREDENTIAL_ALREADY_EXISTS"
        is PrfProviderException.Generic -> "GENERIC_ERROR"
    }
}

File: plugins/capacitor-passkey-prf/ios/Sources/CapacitorPasskeyPrf/PasskeyPrfPlugin.swift
import BreezSdkSpark
import Capacitor
import Foundation

/// Capacitor bridge over `BreezSdkSpark.PasskeyClient`.
///
/// Owns a single `PasskeyClient` per session (built in `initialize`)
/// plus a process-lifetime `KeychainCredentialRegistry`.
@objc(PasskeyPlugin)
public class PasskeyPlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "PasskeyPlugin"
    public let jsName = "Passkey"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "initialize", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "checkAvailability", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "register", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "signIn", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "connectWithPasskey", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "listLabels", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "storeLabel", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "getKnownCredentialIds", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "removeKnownCredentialId", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "clearKnownCredentialIds", returnType: CAPPluginReturnPromise),
    ]

    private let registry: Any? = {
        if #available(iOS 18.0, *) { return KeychainCredentialRegistry() }
        return nil
    }()
    private var client: PasskeyClient?
    /// RP the client was initialized with; needed to key the registry.
    private var rpId: String?

    /// RP IDs this app may run a ceremony against. Must stay in step with
    /// the `webcredentials:` entries in App.entitlements (and Android's
    /// assetlinks.json): a domain here the app is not associated with
    /// dead-ends the OS sheet.
    private static let allowedRpIds: Set<String> = ["keys.hayabusawallet.com"]

    @objc func initialize(_ call: CAPPluginCall) {
        guard #available(iOS 18.0, *) else {
            call.reject("Passkey requires iOS 18.0+", "PRF_NOT_SUPPORTED")
            return
        }
        guard let rpId = call.getString("rpId"), let rpName = call.getString("rpName") else {
            call.reject("rpId and rpName are required", "INVALID_ARGUMENT")
            return
        }
        // The RP is ours to decide, not the WebView's. Settle it here, before
        // a ceremony can put a domain in front of the user.
        guard Self.allowedRpIds.contains(rpId) else {
            call.reject("rpId '\(rpId)' is not an associated domain of this app", "INVALID_ARGUMENT")
            return
        }
        self.rpId = rpId
        let provider = PasskeyProvider(
            options: PasskeyProviderOptions(
                rpId: rpId,
                rpName: rpName,
                userName: call.getString("userName"),
                userDisplayName: call.getString("userDisplayName")
            )
        )
        let config = call.getString("defaultLabel").map { PasskeyConfig(defaultLabel: $0) }
        client = PasskeyClient(prfProvider: provider, breezApiKey: call.getString("breezApiKey"), config: config)
        call.resolve()
    }

    @objc func checkAvailability(_ call: CAPPluginCall) {
        guard let c = client else {
            call.resolve(["type": "prfUnsupported"])
            return
        }
        Task {
            do {
                call.resolve(Self.encodeAvailability(try await c.checkAvailability()))
            } catch {
                call.reject(error.localizedDescription, errorCode(error))
            }
        }
    }

    @objc func register(_ call: CAPPluginCall) {
        guard let c = client else {
            call.reject("Plugin not initialized", "NOT_INITIALIZED")
            return
        }
        let request = RegisterRequest(
            label: call.getString("label"),
            excludeCredentials: parseB64Array(call, key: "excludeCredentials")
        )
        Task {
            do {
                let response = try await c.register(request: request)
                await self.recordCredential(response.credential)
                var out: [String: Any] = ["wallet": Self.encodeWallet(response.wallet)]
                out["credential"] = response.credential.map { Self.encodeCredential($0) } ?? NSNull()
                call.resolve(out)
            } catch {
                call.reject(error.localizedDescription, errorCode(error))
            }
        }
    }

    @objc func signIn(_ call: CAPPluginCall) {
        guard let c = client else {
            call.reject("Plugin not initialized", "NOT_INITIALIZED")
            return
        }
        let request = SignInRequest(
            label: call.getString("label"),
            allowCredentials: parseB64Array(call, key: "allowCredentials"),
            preferImmediatelyAvailableCredentials: call.getBool("preferImmediatelyAvailableCredentials")
        )
        Task {
            do {
                let response = try await c.signIn(request: request)
                await self.recordCredential(response.credential)
                var out: [String: Any] = [
                    "wallet": Self.encodeWallet(response.wallet),
                    "labels": response.labels,
                ]
                out["credentialId"] = response.credential?.credentialId.base64EncodedString() ?? NSNull()
                call.resolve(out)
            } catch {
                call.reject(error.localizedDescription, errorCode(error))
            }
        }
    }

    @objc func connectWithPasskey(_ call: CAPPluginCall) {
        guard let c = client else {
            call.reject("Plugin not initialized", "NOT_INITIALIZED")
            return
        }
        let request = ConnectWithPasskeyRequest(
            label: call.getString("label"),
            allowCredentials: parseB64Array(call, key: "allowCredentials"),
            excludeCredentials: parseB64Array(call, key: "excludeCredentials")
        )
        Task {
            do {
                let response = try await c.connectWithPasskey(request: request)
                await self.recordCredential(response.credential)
                var out: [String: Any] = [
                    "wallet": Self.encodeWallet(response.wallet),
                    // Discovery labels for the returning multi-wallet case;
                    // empty on the register path (glow-web#309).
                    "labels": response.labels,
                ]
                if let cred = response.credential {
                    out["registeredCredential"] = Self.encodeCredential(cred)
                } else {
                    out["registeredCredential"] = NSNull()
                }
                call.resolve(out)
            } catch {
                call.reject(error.localizedDescription, errorCode(error))
            }
        }
    }

    @objc func listLabels(_ call: CAPPluginCall) {
        guard let c = client else {
            call.reject("Plugin not initialized", "NOT_INITIALIZED")
            return
        }
        Task {
            do {
                call.resolve(["labels": try await c.labels().list()])
            } catch {
                call.reject(error.localizedDescription, errorCode(error))
            }
        }
    }

    @objc func storeLabel(_ call: CAPPluginCall) {
        guard let label = call.getString("label") else {
            call.reject("Missing required parameter: label", "INVALID_ARGUMENT")
            return
        }
        guard let c = client else {
            call.reject("Plugin not initialized", "NOT_INITIALIZED")
            return
        }
        Task {
            do {
                try await c.labels().store(label: label)
                call.resolve()
            } catch {
                call.reject(error.localizedDescription, errorCode(error))
            }
        }
    }

    @objc func getKnownCredentialIds(_ call: CAPPluginCall) {
        guard #available(iOS 18.0, *),
              let reg = registry as? KeychainCredentialRegistry,
              let rp = rpId else {
            call.resolve(["credentialIds": [String]()])
            return
        }
        Task {
            do {
                let ids = try await reg.read(rpId: rp)
                call.resolve(["credentialIds": ids.map { $0.base64EncodedString() }])
            } catch {
                call.reject(error.localizedDescription, errorCode(error))
            }
        }
    }

    @objc func removeKnownCredentialId(_ call: CAPPluginCall) {
        guard let idString = call.getString("credentialId"), !idString.isEmpty,
              let id = Data(base64Encoded: idString) else {
            call.reject("Missing or invalid credentialId", "INVALID_ARGUMENT")
            return
        }
        guard #available(iOS 18.0, *),
              let reg = registry as? KeychainCredentialRegistry,
              let rp = rpId else {
            call.resolve()
            return
        }
        Task {
            do {
                try await reg.remove(rpId: rp, credentialId: id)
                call.resolve()
            } catch {
                call.reject(error.localizedDescription, errorCode(error))
            }
        }
    }

    @objc func clearKnownCredentialIds(_ call: CAPPluginCall) {
        guard #available(iOS 18.0, *),
              let reg = registry as? KeychainCredentialRegistry,
              let rp = rpId else {
            call.resolve()
            return
        }
        Task {
            do {
                try await reg.clear(rpId: rp)
                call.resolve()
            } catch {
                call.reject(error.localizedDescription, errorCode(error))
            }
        }
    }

    /// Best-effort: record a credential ID in the synced store.
    private func recordCredential(_ cred: PasskeyCredential?) async {
        guard #available(iOS 18.0, *),
              let cred,
              let reg = registry as? KeychainCredentialRegistry,
              let rp = rpId else { return }
        try? await reg.add(rpId: rp, credentialId: cred.credentialId)
    }

    private func parseB64Array(_ call: CAPPluginCall, key: String) -> [Data] {
        (call.getArray(key) as? [String] ?? []).compactMap { Data(base64Encoded: $0) }
    }

    private static func encodeWallet(_ wallet: Wallet) -> [String: Any] {
        return [
            "seed": encodeSeed(wallet.seed),
            "label": wallet.label,
        ]
    }

    private static func encodeSeed(_ seed: Seed) -> [String: Any] {
        switch seed {
        case .mnemonic(let mnemonic, let passphrase):
            var out: [String: Any] = ["type": "mnemonic", "mnemonic": mnemonic]
            out["passphrase"] = passphrase ?? NSNull()
            return out
        case .entropy(let data):
            return ["type": "entropy", "entropy": data.base64EncodedString()]
        }
    }

    private static func encodeCredential(_ cred: PasskeyCredential) -> [String: Any] {
        var out: [String: Any] = [
            "credentialId": cred.credentialId.base64EncodedString(),
        ]
        out["userId"] = cred.userId?.base64EncodedString() ?? NSNull()
        out["aaguid"] = cred.aaguid?.base64EncodedString() ?? NSNull()
        out["backupEligible"] = cred.backupEligible ?? NSNull()
        return out
    }

    private static func encodeAvailability(_ a: PasskeyAvailability) -> [String: Any] {
        switch a {
        case .available: return ["type": "available"]
        case .prfUnsupported: return ["type": "prfUnsupported"]
        case .notAssociated(let source, let reason):
            return ["type": "notAssociated", "source": source, "reason": reason]
        case .skipped(let reason):
            return ["type": "skipped", "reason": reason]
        }
    }

    private func errorCode(_ error: Error) -> String {
        if let passkey = error as? PasskeyError {
            if case .Prf(let inner) = passkey { return prfErrorCode(inner) }
            switch passkey {
            case .Prf: return "GENERIC_ERROR"  // unreachable; handled above
            case .RelayConnectionFailed: return "RELAY_CONNECTION_FAILED"
            case .NostrWriteFailed: return "NOSTR_WRITE_FAILED"
            case .NostrReadFailed: return "NOSTR_READ_FAILED"
            case .KeyDerivationError: return "KEY_DERIVATION_ERROR"
            case .InvalidPrfOutput: return "INVALID_PRF_OUTPUT"
            case .MnemonicError: return "MNEMONIC_ERROR"
            case .InvalidSalt: return "INVALID_SALT"
            // A passkey exists on the device even though the ceremony
            // failed. Its own code so the web layer signs in with it
            // rather than offering a create that would strand it.
            case .CreatedButNotDerived: return "CREATED_BUT_NOT_DERIVED"
            case .Generic: return "GENERIC_ERROR"
            }
        }
        if let prf = error as? PrfProviderError { return prfErrorCode(prf) }
        return "UNKNOWN_ERROR"
    }

    private func prfErrorCode(_ prf: PrfProviderError) -> String {
        switch prf {
        case .PrfNotSupported: return "PRF_NOT_SUPPORTED"
        case .UserCancelled: return "USER_CANCELLED"
        case .UserTimedOut: return "USER_TIMED_OUT"
        case .CredentialNotFound: return "CREDENTIAL_NOT_FOUND"
        case .AuthenticationFailed: return "AUTHENTICATION_FAILED"
        case .PrfEvaluationFailed: return "PRF_EVALUATION_FAILED"
        case .Configuration: return "CONFIGURATION_ERROR"
        case .CredentialAlreadyExists: return "CREDENTIAL_ALREADY_EXISTS"
        case .Generic: return "GENERIC_ERROR"
        }
    }
}
